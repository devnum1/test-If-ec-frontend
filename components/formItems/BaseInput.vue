// 01EZTX9M3AWK0ZB16B6KYR56F1

<template>
  <div
    :class="[
      'base-input',
      { 'base-input--help': helpText },
      { 'base-input--error': error },
      inputClass,
    ]"
  >
    <div class="base-input__wrapper">
      <input
        :id="id"
        ref="input"
        v-mask="mask"
        :name="name"
        :class="{ 'is-danger': error, tracking: !!id }"
        class="base-input__main form-control"
        :type="inputType"
        :maxlength="maxLength"
        :placeholder="placeholder"
        :value="value"
        :autocomplete="autocomplete"
        :disabled="disabled"
        :min="min"
        @input="handleInput"
        @keyup="handleKeyup"
        @keypress="handleKeyPress"
        @blur="handleBlur"
      />
      <span v-if="error" class="base-input__error">{{ error }}</span>
    </div>
    <span v-if="helpText" class="base-input__text">{{ helpText }}</span>
  </div>
</template>
<script>
export default {
  name: 'BaseInput',
  $_veeValidate: {
    // value getter
    value() {
      return this.$el.value;
    },
    // name getter
    name() {
      return this.name;
    },
  },
  props: {
    value: {
      type: String,
      default: '',
    },
    disabled: {
      type: Boolean,
      default: false,
    },
    name: {
      type: String,
      default: '',
    },
    inputType: {
      type: String,
      default: 'text',
    },
    error: {
      type: String,
      default: '',
    },
    placeholder: {
      type: String,
      default: '',
    },
    inputClass: {
      type: String,
      default: '',
    },
    helpText: {
      type: String,
      default: '',
    },
    autocomplete: {
      type: String,
      default: 'on',
    },
    mask: {
      type: String,
      default: '',
      required: false,
    },
    id: {
      type: String,
      default: '',
    },
    maxLength: {
      type: String,
      default: '',
    },
    min: {
      type: String,
      default: '',
    },
  },
  data() {
    return {
      isComposing: false,
    };
  },
  methods: {
    handleKeyup(e) {
      this.$emit('keyup', e);
    },
    handleKeyPress(e) {
      this.$emit('keypress', e);
    },
    handleInput(e) {
      this.$emit('input', e.target.value);
    },
    handleBlur(e) {
      this.$emit('blur', e);
    },
  },
};
</script>
<style lang="scss">
@import '~/assets/scss/components/base-inputs';
</style>
