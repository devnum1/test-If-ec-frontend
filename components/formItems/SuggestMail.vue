// 01EZTX9M3AWK0ZB16B6KYR56F1

<template>
  <div v-click-outside="onClickOutside" class="email-dropdown">
    <div class="email-dropdown__input">
      <BaseInput
        :id="id"
        v-model="email"
        v-validate="{ required: true, email: ['XXX@XXX.XXX'] }"
        :data-vv-as="$t('common.mail')"
        name="email"
        :value="email"
        :error="errors.first('email')"
        :placeholder="placeholderMail"
        autocomplete="email"
        @keyup="handleEnter"
      />
    </div>
    <div
      :class="{
        'email-dropdown__container': true,
        'd-none': hideList,
      }"
    >
      <ul v-if="shouldShowList" class="email-dropdown__list">
        <li
          v-for="(domain, index) in domainsList"
          :key="index"
          tabindex="-1"
          :data-dropdown-item-index="index"
          class="email-dropdown__item"
          @click="handleOptionSelection(domain)"
          @keyup.enter="handleOptionSelection(domain)"
        >
          <span class="email-dropdown__item-local-part">{{ localPart }}</span>
          <span class="email-dropdown__item-domain">@{{ domain }}</span>
        </li>
      </ul>
    </div>
  </div>
</template>

<script type="text/javascript">
import vClickOutside from 'v-click-outside';
import VeeValidate from 'vee-validate';
import BaseInput from '@/components/formItems/BaseInput';
import VueI18n from 'vue-i18n';

export default {
  components: { BaseInput },
  directives: {
    clickOutside: vClickOutside.directive,
  },
  props: {
    id: {
      type: String,
      default: '',
    },
    value: {
      type: String,
      default: '',
    },
    defaultDomains: {
      type: Array,
      default() {
        return ['gmail.com', 'yahoo.co.jp', 'outlook.com', 'icloud.com'];
      },
    },
    maxSuggestions: {
      type: Number,
      default: 7,
    },
    isActive: {
      type: Boolean,
      default: false,
    },
    placeholderMail: {
      type: String,
      default: 'ä¾‹ ) peco@hmail.com',
    },
  },

  data() {
    return {
      email: this.value,
      hideList: true,
    };
  },

  computed: {
    shouldShowList() {
      return this.domainsList.length && !this.isOptionSelected;
    },
    localPart() {
      return String(this.email).toLowerCase().split('@')[0];
    },
    inputDomain() {
      const parts = String(this.email).toLowerCase().split('@');

      return parts.length > 1 ? parts[1] : '';
    },
    suggestionList() {
      return this.domainsList.map((domain) =>
        `${this.localPart}@${domain}`.toLowerCase()
      );
    },
    isOptionSelected() {
      return this.suggestionList.includes(String(this.email).toLowerCase());
    },
    domainsList() {
      return this.defaultDomains
        .filter((domain) => domain.indexOf(this.inputDomain) === 0)
        .slice(0, this.maxSuggestions);
    },
  },

  watch: {
    email: function (newValue) {
      if (/@./.test(newValue)) {
        this.hideList = false;
      } else {
        this.hideList = true;
      }

      this.$emit('input', newValue);
    },
    isActive: function () {
      this.handleValidate();
    },
  },

  methods: {
    handleOptionSelection(domain) {
      this.email = `${this.localPart}@${domain}`;
    },
    onClickOutside(event) {
      this.hideList = true;
    },
    handleEnter(e) {
      if (e.key === 'Enter') {
        this.hideList = true;
      }
    },
    handleValidate() {
      this.$validator.validate().then((result) => {
        //
      });
    },
  },
};
</script>

<style lang="scss" scoped>
@import '~/assets/scss/components/suggest-mail';
</style>
