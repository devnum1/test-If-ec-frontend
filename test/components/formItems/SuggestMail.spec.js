import { shallowMount, createLocalVue } from '@vue/test-utils';
import SuggestMail from '@/components/formItems/SuggestMail';
import BaseInput from '@/components/formItems/BaseInput';
import VeeValidate from 'vee-validate';
import vClickOutside from 'v-click-outside';

const $t = () => {};
let wrapper;
const localVue = createLocalVue();
localVue.use(BaseInput);
localVue.use(VeeValidate);

const factory = () => {
  return shallowMount(SuggestMail, {
    propsData: {
      name: 'test',
    },
    mocks: {
      $t,
      email: 'peco@hmail.com',
    },
    stubs: { 'base-input': BaseInput },
    methods: {},
    localVue,
    directives: {
      vClickOutside,
      clickOutside: vClickOutside.directive,
    },
  });
};

beforeEach(() => {
  wrapper = factory();
});

afterEach(() => {
  wrapper.destroy();
});

jest.useFakeTimers();

describe('SuggestMail', () => {
  const spySelection = jest.spyOn(SuggestMail.methods, 'handleOptionSelection');

  test('is a Vue instance', () => {
    expect(wrapper.vm).toBeTruthy();
  });

  test('match the snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  test('emits input event when email changes', () => {
    wrapper.vm.$options.watch.email.call(wrapper.vm, 'peco@hmail.com');
    expect(wrapper.emitted('input')[0][0]).toBe('peco@hmail.com');
  });

  test('emits input event when email changes hide list', () => {
    wrapper.vm.$options.watch.email.call(wrapper.vm, 'peco');
    expect(wrapper.emitted('input')[0][0]).toBe('peco');
  });

  test('Check validate when input email', () => {
    wrapper.vm.handleValidate = jest.fn();
    wrapper.vm.$options.watch.isActive.call(wrapper.vm, true);
    expect(wrapper.vm.handleValidate).toHaveBeenCalled();
  });

  test('Test handle option select function', async () => {
    const emailButton = wrapper.find('.email-dropdown__item');
    await emailButton.trigger('click');

    expect(spySelection).toHaveBeenCalled();
  });
});
