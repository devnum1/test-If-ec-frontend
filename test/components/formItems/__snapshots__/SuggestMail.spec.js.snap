// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SuggestMail match the snapshot 1`] = `
"<div class=\\"email-dropdown\\" name=\\"test\\">
  <div class=\\"email-dropdown__input\\">
    <baseinput-stub value=\\"\\" name=\\"email\\" inputtype=\\"text\\" error=\\"\\" placeholder=\\"ä¾‹ ) peco@hmail.com\\" inputclass=\\"\\" helptext=\\"\\" autocomplete=\\"email\\" mask=\\"\\" id=\\"\\" maxlength=\\"\\" min=\\"\\" aria-required=\\"true\\" aria-invalid=\\"false\\"></baseinput-stub>
  </div>
  <div class=\\"email-dropdown__container d-none\\">
    <ul class=\\"email-dropdown__list\\">
      <li tabindex=\\"-1\\" data-dropdown-item-index=\\"0\\" class=\\"email-dropdown__item\\"><span class=\\"email-dropdown__item-local-part\\"></span> <span class=\\"email-dropdown__item-domain\\">@gmail.com</span></li>
      <li tabindex=\\"-1\\" data-dropdown-item-index=\\"1\\" class=\\"email-dropdown__item\\"><span class=\\"email-dropdown__item-local-part\\"></span> <span class=\\"email-dropdown__item-domain\\">@yahoo.co.jp</span></li>
      <li tabindex=\\"-1\\" data-dropdown-item-index=\\"2\\" class=\\"email-dropdown__item\\"><span class=\\"email-dropdown__item-local-part\\"></span> <span class=\\"email-dropdown__item-domain\\">@outlook.com</span></li>
      <li tabindex=\\"-1\\" data-dropdown-item-index=\\"3\\" class=\\"email-dropdown__item\\"><span class=\\"email-dropdown__item-local-part\\"></span> <span class=\\"email-dropdown__item-domain\\">@icloud.com</span></li>
    </ul>
  </div>
</div>"
`;
